sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCast.scala

```diff
@@ -938,8 +916,35 @@ object GpuCast extends Arm {
           convertFixedLenDateOrNull(sanitizedInput, 4, "%Y")))
 
       // handle special dates like "epoch", "now", etc.
-      specialDates.foldLeft(converted)((prev, specialDate) =>
-        specialDateOr(sanitizedInput, specialDate._1, specialDate._2, prev))
+      withResource(daysEqual(sanitizedInput, DateUtils.EPOCH)) { isEpoch =>
```

Comment:
This change feels like a step backwards.  Before with the folding the maximum extra memory needed on the GPU was a boolean column, and the new output column.  With this change we now have 5 boolean columns and 5 temporary output columns.

Reply:
Yes, it was not optimal.  I reworked this part, cleaning up temporary GPU resources as early as possible.